src/installer/tests/TestUtils/SingleFileTestApp.cs

```diff
@@ -0,0 +1,155 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Reflection;
+using Microsoft.NET.HostModel.Bundle;
+
+namespace Microsoft.DotNet.CoreSetup.Test
+{
+    public class SingleFileTestApp : TestArtifact
+    {
+        public string AppName { get; }
+        public string NonBundledLocation => builtApp.Location;
+
+        private readonly TestApp builtApp;
+        private readonly bool selfContained;
+
+        public SingleFileTestApp(string appName, bool selfContained, string location)
+            : base(location)
+        {
+            AppName = appName;
+            this.selfContained = selfContained;
+
+            builtApp = new TestApp(Path.Combine(Location, "builtApp"), AppName);
+            Directory.CreateDirectory(builtApp.Location);
+            PopulateBuiltAppDirectory();
+        }
+
+        public static SingleFileTestApp CreateFrameworkDependent(string appName)
+            => Create(appName, selfContained: false);
+
+        public static SingleFileTestApp CreateSelfContained(string appName)
+            => Create(appName, selfContained: true);
+
+        private static SingleFileTestApp Create(string appName, bool selfContained)
+        {
+            var (location, parentPath) = GetNewTestArtifactPath(appName);
+            return new SingleFileTestApp(appName, selfContained, location)
+            {
+                DirectoryToDelete = parentPath
+            };
+        }
+
+        public string Bundle(BundleOptions options, Version? bundleVersion = null)
+        {
+            string bundleDirectory = SharedFramework.CalculateUniqueTestDirectory(Path.Combine(Location, "bundle"));
+            var bundler = new Bundler(
+                Binaries.GetExeFileNameForCurrentPlatform(AppName),
+                bundleDirectory,
+                options,
+                targetFrameworkVersion: bundleVersion,
+                macosCodesign: true);
+
+            // Get all files in the source directory and all sub-directories.
+            string[] sources = Directory.GetFiles(builtApp.Location, searchPattern: "*", searchOption: SearchOption.AllDirectories);
+
+            // Sort the file names to keep the bundle construction deterministic.
+            Array.Sort(sources, StringComparer.Ordinal);
```

Comment:
Nit - would it make more sense to sort the final `fileSpecs` right before GenerateBundle. The self-contained construction might in theory produce different order of files as well.